File: site/js/hydrology.js
/**
 * The ManyFEWS catchment model in JavaScript.
 *
 * A direct transcription of manyfews_core/hydrology.py, which is itself a
 * verbatim port of the Django implementation, which is itself a port of the
 * original MATLAB. The variable names are kept deliberately unidiomatic so the
 * three versions can be diffed against each other line by line.
 *
 * Parity with the MATLAB reference outputs in Data/*_Benchmark.csv is checked by
 * site/test/parity.test.mjs. If you change anything numeric here, run it.
 *
 * References:
 *   Allen et al. (1998), FAO Irrigation and Drainage Paper 56.
 *   Moore (2007), The PDM rainfall-runoff model, HESS 11(1), 483-499.
 *   Mathias et al. (2016), J. Hydrol. 540, 423-436.
 */

/** Surface roughness equivalent to the FAO56 reference crop. */
const Z0 = 0.006247;

/**
 * FAO56 Penman-Monteith reference evapotranspiration.
 *
 * @param {number} dt time step in days; must be 0.25 (see the day-of-year note)
 * @param {Date} predictionDate start of the series, for day-of-year
 * @param {Float64Array} Tmin daily minimum temperature, degrees C, per bucket
 * @param {Float64Array} Tmax daily maximum temperature, degrees C, per bucket
 * @param {number} alt catchment mean altitude, m
 * @param {number} lat catchment mean latitude, degrees
 * @param {Float64Array} T air temperature, degrees C
 * @param {Float64Array} u2 wind speed at 2 m, m/s
 * @param {Float64Array} RH relative humidity, %
 * @returns {{ETo: Float64Array, E0: Float64Array}} mm/day
 */
export function fao56(dt, predictionDate, Tmin, Tmax, alt, lat, T, u2, RH) {
  const n = Tmax.length;

  // UPSTREAM QUIRK, PRESERVED: the original reassigns Tmax and then derives
  // Tmin from the *new* value, so the intended swap-guard does nothing at all.
  // Reproduced here because parity depends on it.
  const tmax = new Float64Array(n);
  const tmin = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    tmax[i] = Math.max(Tmax[i], Tmin[i]);
    tmin[i] = Math.min(tmax[i], Tmin[i]);
  }

  // Atmospheric pressure, Eq. 7, and psychrometric constant, Eq. 8.
  const P = 101.3 * Math.pow((293 - 0.0065 * alt) / 293, 5.26);
  const gam = ((1.013e-3 * P) / 0.622) / 2.45;

  const varphi = (lat * Math.PI) / 180;
  const startOfYear = Date.UTC(predictionDate.getUTCFullYear() - 1, 11, 31);
  const dayOfYear = Math.round(
    (Date.UTC(
      predictionDate.getUTCFullYear(),
      predictionDate.getUTCMonth(),
      predictionDate.getUTCDate()
    ) - startOfYear) / 86400000
  );

  const ETo = new Float64Array(n);
  const E0 = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    const t = T[i];

    // Slope of the saturation vapour pressure curve, Eq. 13.
    const Del = (4098 * (0.6108 * Math.exp((17.27 * t) / (t + 237.3)))) / ((t + 237.3) ** 2);

    // Saturation and actual vapour pressure, Eqs. 11 and 14.
    const eoTmax = 0.6108 * Math.exp((17.27 * tmax[i]) / (tmax[i] + 237.3));
    const eoTmin = 0.6108 * Math.exp((17.27 * tmin[i]) / (tmin[i] + 237.3));
    const es = (eoTmax + eoTmin) / 2;
    const ea = (RH[i] / 100) * es;

    // Day-of-year advances by dt per bucket - only correct at four per day.
    const J = dayOfYear + i * dt;

    const dr = 1 + 0.033 * Math.cos(((2 * Math.PI) / 365) * J);          // Eq. 23
    const delta = 0.409 * Math.sin(((2 * Math.PI) / 365) * J - 1.39);    // Eq. 24
    const ws = Math.acos(-Math.tan(varphi) * Math.tan(delta));           // Eq. 25

    // Extraterrestrial radiation, Eq. 21.
    const Ra =
      ((24 * 60) / Math.PI) * 0.082 * dr *
      (ws * Math.sin(varphi) * Math.sin(delta) +
        Math.cos(varphi) * Math.cos(delta) * Math.sin(ws));

    // Incoming solar radiation from the diurnal range, Eq. 50 (Hargreaves).
    const Rs = 0.16 * Math.sqrt(tmax[i] - tmin[i]) * Ra;
    const Rso = (0.75 + 2e-5 * alt) * Ra;                                // Eq. 37
    const RsRso = Math.min(Rs / Rso, 1);

    // Net longwave, Eq. 39.
    const sig = 4.903e-9;
    const sigT4 =
      (sig * (tmax[i] + 273.15) ** 4 + sig * (tmin[i] + 273.15) ** 4) / 2;
    const Rnl = sigT4 * (0.34 - 0.14 * Math.sqrt(ea)) * (1.35 * RsRso - 0.35);

    // Reference crop: albedo 0.23, Eqs. 38 and 6.
    const T2 = ((gam * 900) / (t + 273)) * u2[i] * (es - ea);
    let Rn = 0.77 * Rs - Rnl;
    ETo[i] = (0.408 * Del * Rn + T2) / (Del + gam * (1 + 0.34 * u2[i]));

    // Open water: albedo 0.05 and no surface resistance. T2 carries over
    // unchanged - it has no albedo term.
    Rn = 0.95 * Rs - Rnl;
    E0[i] = (0.408 * Del * Rn + T2) / (Del + gam);
  }

  return { ETo, E0 };
}

/**
 * Probability Distributed Model soil store (Moore 2007).
 *
 * @returns {{qro: Float64Array, qd: Float64Array, S: Float64Array}}
 *   surface runoff and drainage in mm/day, storage in mm
 */
export function pdmModel(qp, Ep, Smax, gamma, k, dt, S0) {
  const n = qp.length;
  const qd = new Float64Array(n);
  const qro = new Float64Array(n);
  const S = new Float64Array(n + 1);
  S[0] = S0;

  for (let i = 0; i < n; i++) {
    // With gamma pinned to 1 by the caller this is simply S/Smax.
    const F = 1 - Math.pow(1 - S[i] / Smax, gamma);
    qd[i] = (k * S[i]) / Smax;

    const Strial = S[i] + ((1 - F) * qp[i] - Ep[i] - qd[i]) * dt;
    S[i + 1] = Strial;
    qro[i] = F * qp[i];

    if (Strial <= 0) {
      S[i + 1] = 0;
      qd[i] = 0;
    } else if (Strial >= Smax) {
      S[i + 1] = Smax;
      qro[i] = qp[i] - Ep[i] - (Smax - S[i]) / dt - qd[i];
    }
  }

  return { qro, qd, S: S.subarray(0, n) };
}

/**
 * Non-linear routing store, q = a * v^b (Mathias et al. 2016).
 *
 * @param {number} X residence time in days when b === 1, else qmax in mm/day
 */
export function routingFun(qs, X, b, dt, q0) {
  const n = qs.length;
  let a;
  let vmax;

  if (b === 1) {
    a = 1 / X;                       // X is the residence time
    vmax = Infinity;
  } else {
    const dtDAY = 1;                 // qmax was determined from daily data
    a = Math.pow(X, 1 - b) * Math.pow(b * dtDAY, -b);
    vmax = Math.pow(a * b * dt, 1 / (1 - b));
  }

  const q = new Float64Array(n + 1).fill(q0);
  const v = new Float64Array(n + 1);
  v[0] = Math.pow(q0 / a, 1 / b);

  for (let i = 0; i < n; i++) {
    const qtrial = a * Math.pow(v[i], b);
    const vtrial = v[i] + (qs[i] - qtrial) * dt;

    if (vtrial < vmax) {
      q[i] = qtrial;
      v[i + 1] = vtrial;
    } else {
      q[i] = qs[i] - (vmax - v[i]) / dt;
      v[i + 1] = vmax;
    }
  }

  return q.subarray(0, n);
}

/**
 * Run every parameter set: PDM, then the slow and fast routing stores.
 *
 * @param {Float64Array} qp rainfall, mm/day
 * @param {Float64Array} Ep potential evapotranspiration, mm/day
 * @param {Array<Array<number>>} X parameter sets [Smax, qmax, k, Tr]
 * @param {Array<Float64Array>} F0 per-set state [storage, slow, fast];
 *   **mutated in place**, exactly as the Python and MATLAB versions do
 * @returns {Array<Float64Array>} flow in m3/s, one array per parameter set
 */
export function modelFun(qp, Ep, dt, catAreaKm2, X, F0) {
  const toCumecs = (catAreaKm2 * 1e3) / 24 / 3600;
  const flows = [];

  for (let s = 0; s < X.length; s++) {
    const [Smax, qmax, k, Tr] = X[s];
    const [S0, qSlow0, qFast0] = F0[s];

    // gamma is pinned to 1 upstream despite the Pareto docstring.
    const { qro, qd, S } = pdmModel(qp, Ep, Smax, 1, k, dt, S0);
    const qSlow = routingFun(qd, Tr, 1, dt, qSlow0);
    const qFast = routingFun(qro, qmax, 5 / 3, dt, qFast0);

    const q = new Float64Array(qp.length);
    for (let i = 0; i < qp.length; i++) q[i] = (qSlow[i] + qFast[i]) * toCumecs;
    flows.push(q);

    F0[s] = [S[S.length - 1], qSlow[qSlow.length - 1], qFast[qFast.length - 1]];
  }

  return flows;
}

/** Log-profile conversion of 10 m wind magnitude to 2 m. */
export function wind10mTo2m(u, v) {
  const u10 = Math.hypot(u, v);
  const uTau = u10 / 2.5 / Math.log(10 / Z0);
  return 2.5 * uTau * Math.log(2 / Z0);
}

/**
 * The whole catchment model for one weather realisation.
 *
 * @param {{times: Date[], rh, tmaxK, tminK, windU, windV, precip}} weather
 *   six-hourly, chronological, a whole number of days
 * @param {Array<Array<number>>} params 100 parameter sets
 * @param {Array<Array<number>>} state per-set initial state; not mutated
 * @param {{latitude_deg, altitude_m, area_km2, timestep_days}} catchment
 * @returns {{flows: Array<Float64Array>, state: Array, rainfall, pet}}
 */
export function generateRiverFlows(weather, params, state, catchment) {
  const dt = catchment.timestep_days;
  const n = weather.precip.length;
  if (n % 4 !== 0) {
    throw new Error(
      `weather has ${n} buckets, not a whole number of days; the daily ` +
        'temperature grouping assumes four 6-hour buckets per day'
    );
  }

  const tempMax = new Float64Array(n);
  const tempMin = new Float64Array(n);
  const T = new Float64Array(n);
  const u2 = new Float64Array(n);
  const qp = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    tempMax[i] = weather.tmaxK[i] - 273.15;
    tempMin[i] = weather.tminK[i] - 273.15;
    T[i] = (tempMin[i] + tempMax[i]) / 2;
    u2[i] = wind10mTo2m(weather.windU[i], weather.windV[i]);
    qp[i] = weather.precip[i] / dt;
  }

  // Collapse each day's four buckets to one min and one max, then broadcast back.
  const dailyMin = new Float64Array(n);
  const dailyMax = new Float64Array(n);
  for (let d = 0; d < n / 4; d++) {
    let lo = Infinity;
    let hi = -Infinity;
    for (let j = 0; j < 4; j++) {
      lo = Math.min(lo, tempMin[d * 4 + j]);
      hi = Math.max(hi, tempMax[d * 4 + j]);
    }
    for (let j = 0; j < 4; j++) {
      dailyMin[d * 4 + j] = lo;
      dailyMax[d * 4 + j] = hi;
    }
  }

  const { ETo } = fao56(
    dt, weather.times[0], dailyMin, dailyMax,
    catchment.altitude_m, catchment.latitude_deg, T, u2, weather.rh
  );

  // modelFun writes end state back into what it is given, so hand it a copy.
  const working = state.map((row) => [row[0], row[1], row[2]]);
  const flows = modelFun(qp, ETo, dt, catchment.area_km2, params, working);

  return { flows, state: working, rainfall: qp, pet: ETo };
}

File: core/manyfews_core/hydrology.py
"""
Verbatim numeric port of ``manyfews/calculations/generate_river_flows.py``
lines 20-350: ``ModelFun``, ``RoutingFun``, ``PDMmodel`` and ``FAO56``.

DO NOT REFACTOR. The CamelCase names, the argument orders and several genuine
oddities are preserved exactly so that a side-by-side diff against the Django
module stays reviewable, and so that the MATLAB reference outputs in
``Data/*_Benchmark.csv`` keep passing. Parity is asserted by
``tests/test_hydrology_benchmark.py``.

Removed relative to the original, none of which affect the numerics:

* ``from django.conf import settings`` and ``from .models import ...`` - neither
  is referenced by any of these four functions.
* The celery task logger, replaced by a stdlib logger.
* Five dead ``try: <name> / except NameError:`` blocks. In each case the name is
  a function parameter, so the bare lookup always succeeds and the ``except``
  body is unreachable: ``q0`` (RoutingFun), ``S0`` (PDMmodel), and ``T``, ``u2``
  and the ``ea`` fallback (FAO56).

A sixth block that *looks* identical is not. See the note above ``Rs`` in
``FAO56``.

Science references:

* Allen, R. G., Pereira, L. S., Raes, D., & Smith, M. (1998). FAO Irrigation and
  Drainage Paper No. 56. FAO.
* Moore, R. J. (2007). The PDM rainfall-runoff model. Hydrology and Earth System
  Sciences, 11(1), 483-499.
* Mathias, S. A., McIntyre, N., & Oughton, R. H. (2016). A study of non-linearity
  in rainfall-runoff response using 120 UK catchments. Journal of Hydrology, 540,
  423-436.
"""

from __future__ import annotations

import logging
import math
from datetime import date, datetime

import numpy as np

logger = logging.getLogger(__name__)

__all__ = ["ModelFun", "RoutingFun", "PDMmodel", "FAO56"]


def ModelFun(
    qp: np.ndarray,
    Ep: np.ndarray,
    dt: float,
    CatArea: float,
    X: np.ndarray,
    F0: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    """
    Determine river flow from weather data by running PDM plus two routing stores
    once per parameter set.

    :param qp: rainfall (mm/day), shape (n_steps,)
    :param Ep: potential evapotranspiration (mm/day), shape (n_steps,)
    :param dt: time step (day)
    :param CatArea: catchment area (km2)
    :param X: model parameters, shape (n_sets, 4) - Smax, qmax, k, Tr
    :param F0: initial state, shape (n_sets, 3) - storage, slow flow, fast flow

    :return: ``(Q, F0)`` where Q is river flow (m3/s) with shape
        (n_steps, n_sets), and F0 is the end-of-run state.

    .. warning::
       ``F0`` is **mutated in place** as well as returned. This matches the
       original and callers rely on it, but it means every caller must pass a
       copy if it intends to reuse the starting state - which is exactly what
       running several ensemble members from one spin-up requires.
    """
    Q = np.zeros(((np.size(Ep[:])), (np.size(X[:, 1]))))  # initialize the matrix Q

    for n in range(np.size(X[:, 1])):
        # Extract parameters
        Smax = X[n, 0]  # (mm)
        qmax = X[n, 1]  # (mm/day)
        k = X[n, 2]  # (mm/day)
        Tr = X[n, 3]  # (days)

        # Extract initial conditions
        S0 = F0[n, 0]  # initial storage level for PDM (mm)
        qSLOW0 = F0[n, 1]  # initial slow flow rate (mm/day)
        qFAST0 = F0[n, 2]  # initial fast flow rate (mm/day)

        # Determine surface runoff and drainage.
        # UPSTREAM QUIRK, PRESERVED: gamma is pinned to 1 despite PDMmodel's
        # Pareto docstring, which makes the store's fraction-saturated term
        # F = S / Smax linear rather than Pareto-distributed.
        qro, qd, Ea, S = PDMmodel(qp, Ep, Smax, 1, k, dt, S0)

        # Determine slow flow (linear store, X is residence time in days)
        qSLOW = RoutingFun(qd, Tr, 1, dt, qSLOW0)

        # Determine fast flow (non-linear store, X is qmax in mm/day)
        qFAST = RoutingFun(qro, qmax, 5 / 3, dt, qFAST0)

        # Determine river flow
        q = qFAST + qSLOW

        # Convert mm/day over the catchment to m3/s
        Q[:, n] = (q * CatArea * (1e3) / 24) / (3600)

        # Update initial condition vector with final values of state variables
        F0[n, :] = [(S[-1]), (qSLOW[-1]), (qFAST[-1])]

    return Q, F0


def RoutingFun(
    qs: np.ndarray, X: float, b: float, dt: float, q0: float
) -> np.ndarray:
    """
    Route runoff through a non-linear store, ``q = a * v**b``, after
    Mathias et al. (2016).

    :param qs: inflow rate (mm/day)
    :param X: residence time in days when ``b == 1``, else qmax in mm/day
    :param b: exponent; 1 gives a linear store, 5/3 the non-linear fast store
    :param dt: time step (day)
    :param q0: initial river flow (mm/day)

    :return: river flow (mm/day), same length as ``qs``
    """
    numPoint = np.size(qs)  # Determine number of data points

    if b == 1:
        # This means it's a linear store
        # so X is the residence time in days
        Tr = X
        a = 1 / Tr
        vmax = float("inf")

    else:
        # This means it's a non-linear store
        # so X is qmax in mm/day
        qmax = X
        dtDAY = 1  # this is needed because qmax is determined with daily data
        a = (math.pow(qmax, (1 - b))) * (math.pow((b * dtDAY), (-b)))
        vmax = math.pow((a * b * dt), (1 / (1 - b)))  # Limit on v for stability

    # UPSTREAM QUIRK, PRESERVED: q is filled with the constant q0 and v derived
    # from all n+1 entries, but only v[0] survives the loop below. Kept because
    # it also sets how vmax interacts with the very first step.
    q = np.full((numPoint + 1), q0)
    v = np.power((q / a), (1 / b))

    for i in range(numPoint):  # Step through each time step
        # Trial values for q and v:
        qtrial = a * math.pow(v[i], b)
        vtrial = v[i] + ((qs[i] - qtrial) * dt)

        if vtrial < vmax:  # Ordinarily use trial values
            q[i] = qtrial  # River flow (mm/day)
            v[i + 1] = vtrial  # River storage (mm)
        else:  # Force v<=vmax
            q[i] = qs[i] - ((vmax - v[i]) / dt)
            v[i + 1] = vmax

    q = q[:-1]

    return q


def PDMmodel(
    qp: np.ndarray,
    Ep: np.ndarray,
    Smax: float,
    gamma: float,
    k: float,
    dt: float,
    S0: float,
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Probability Distributed Model store with a Pareto distribution of store
    capacity, after Moore (2007).

    :param qp: rainfall (mm/day)
    :param Ep: potential evapotranspiration (mm/day)
    :param Smax: maximum storage (mm)
    :param gamma: exponent for the Pareto distribution
    :param k: drainage parameter (mm/day)
    :param dt: time step (day)
    :param S0: initial storage (mm)

    :return: ``(qro, qd, Ea, S)`` - surface runoff, drainage, actual
        evapotranspiration (all mm/day) and catchment storage (mm)
    """
    numPoint = np.size(qp)

    # Initialise vectors
    qd = np.zeros(numPoint)
    qro = np.zeros(numPoint)
    Ea = np.zeros(numPoint)

    S = np.full((numPoint + 1), S0)

    for i in range(numPoint):
        # Pareto CDF
        F = 1 - (math.pow((1 - (S[i]) / Smax), gamma))

        # Determine drainage rate
        qd[i] = k * S[i] / Smax

        # Trial value for S
        Strial = S[i] + (((1 - F) * qp[i] - Ep[i] - qd[i]) * dt)

        # To start with try the following:
        S[i + 1] = Strial  # Catchment storage
        qro[i] = F * qp[i]  # River flow contribution
        Ea[i] = Ep[i]  # Actual evapotranspiration

        if Strial <= 0:
            S[i + 1] = 0
            qd[i] = 0
            Ea[i] = ((1 - F) * qp[i]) + (S[i] / dt)
        elif Strial >= Smax:  # Force S<=Smax
            S[i + 1] = Smax
            qro[i] = qp[i] - Ep[i] - ((Smax - S[i]) / dt) - qd[i]

    S = S[:-1]

    return qro, qd, Ea, S


def FAO56(
    dt: float,
    predictionDate: datetime,
    Tmin: np.ndarray,
    Tmax: np.ndarray,
    alt: float,
    lat: float,
    T: np.ndarray,
    u2: np.ndarray,
    RH: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    """
    FAO56 Penman-Monteith reference evapotranspiration, with open-water
    evaporation as a by-product.

    :param dt: time step (day); see the note on day-of-year below
    :param predictionDate: start of the series, used for day-of-year
    :param Tmin: daily minimum air temperature (deg C)
    :param Tmax: daily maximum air temperature (deg C)
    :param alt: catchment mean altitude (m above sea level)
    :param lat: catchment mean latitude (degrees)
    :param T: air temperature (deg C)
    :param u2: wind speed at 2 m (m/s)
    :param RH: relative humidity (%)

    :return: ``(ETo, E0)`` - reference evapotranspiration and open-water
        evaporation, both mm/day
    """
    # UPSTREAM QUIRK, PRESERVED: the second line reads the Tmax that the first
    # line just reassigned, so `np.minimum(Tmax, Tmin)` is Tmin and the whole
    # statement is a no-op. Fixing it would change the numerics and break parity.
    Tmax = np.maximum(Tmax, Tmin)
    Tmin = np.minimum(Tmax, Tmin)

    # u2 (m/s), P (kPa), ea (kPa), Rn (MJ/m2/day)

    # Slope of saturation curve (Del) from Eq. 13
    Del = (4098 * (0.6108 * (np.exp(((17.27 * T) / (T + 237.3)))))) / np.square(
        T + 237.3
    )

    # Atmospheric pressure (P) from Eq. 7
    P = 101.3 * (math.pow(((293 - 0.0065 * alt) / 293), 5.26))

    # Psychrometric constant (gam) from Eq. 8
    cp = 1.013e-3
    lam = 2.45
    eps = 0.622
    gam = ((cp * P) / eps) / lam

    # Saturation vapour pressure (eo) at Tmax and Tmin from Eq. 11
    eoTmax = 0.6108 * (np.exp((17.27 * Tmax) / (Tmax + 237.3)))
    eoTmin = 0.6108 * (np.exp((17.27 * Tmin) / (Tmin + 237.3)))

    # Assume saturation vapour pressure is mean
    es = (eoTmax + eoTmin) / 2

    # Actual vapour pressure from relative humidity. The original wrapped this in
    # a try/except NameError whose fallback (dewpoint from Tmin) was unreachable,
    # because RH is a parameter.
    ea = (RH / 100) * es

    # Convert latitude from degrees to radians from Eq. 22
    varphi = (lat * math.pi) / 180

    # Determine day of the year as a number from 1 to 365.
    # NOTE: `np.size(Tmax) / 4` assumes four buckets per day, so this only yields
    # one J per input row when dt == 0.25. CatchmentConfig enforces that.
    beginDate = predictionDate.date()
    beginDateNum = (beginDate - date(beginDate.year - 1, 12, 31)).days
    J = beginDateNum + np.arange(0, ((np.size(Tmax[:])) / 4), dt)

    # Inverse relative distance Earth-Sun from Eq. 23
    dr = 1 + (0.033 * np.cos(((2 * math.pi) / 365) * J))

    # Solar declination from Eq. 24
    delta = 0.409 * np.sin((((2 * math.pi) / 365) * J) - 1.39)

    # Sunset hour angle from Eq. 25
    ws = np.arccos((-math.tan(varphi)) * (np.tan(delta)))

    # Extraterrestrial radiation from Eq. 21
    Gsc = 0.0820
    Ra = (
        (((24 * 60) / (math.pi)) * Gsc)
        * dr
        * (
            ws * (math.sin(varphi)) * (np.sin(delta))
            + (math.cos(varphi)) * (np.cos(delta)) * np.sin(ws)
        )
    )

    # Incoming solar radiation from Eq. 50 (Hargreaves radiation formula).
    #
    # CAREFUL. In the original this sat inside `try: Rs / except NameError:`,
    # which looks like the five dead blocks removed elsewhere in this module but
    # is the exact opposite: `Rs` is not a parameter and is never bound earlier,
    # so the bare lookup ALWAYS raises and the except body is the LIVE path.
    # Deleting this line "as dead code" makes Rso/RsRso below raise NameError.
    # tests/test_fao56.py::test_rs_assignment_is_live guards it.
    kRS = 0.16
    Rs = kRS * (np.sqrt(Tmax - Tmin)) * Ra

    # Clear-sky solar radiation from Eq. 37
    Rso = ((0.75 + (2e-5) * alt)) * Ra

    # Net solar radiation from Eq. 38
    alpha = 0.23
    Rns = (1 - alpha) * Rs

    # Outgoing net longwave radiation from Eq. 39
    sig = 4.903e-9
    sigTmax4 = sig * (np.power((Tmax + 273.15), 4))
    sigTmin4 = sig * (np.power((Tmin + 273.15), 4))
    sigT4 = (sigTmax4 + sigTmin4) / 2
    RsRso = Rs / Rso
    RsRso[RsRso > 1] = 1
    Rnl = sigT4 * (0.34 - (0.14 * np.sqrt(ea))) * (1.35 * RsRso - 0.35)

    # For the UK you would instead want:
    # Rnl = sigT4 * (0.56 - (0.25 * np.sqrt(ea))) * (1.35 * RsRso - 0.35)

    # Net radiation from Eq. 40
    Rn = Rns - Rnl

    # Assume zero soil heat flux
    G = 0

    # Calculate reference evapotranspiration from Eq. 6
    T1 = 0.408 * Del * (Rn - G)
    T2 = ((gam * 900) / (T + 273)) * u2 * (es - ea)
    T3 = Del + (gam * (1 + 0.34 * u2))
    ETo = (T1 + T2) / T3

    # Determine open water evaporation. T2 is deliberately not recomputed - it
    # has no albedo term - while T1 picks up the new net radiation and T3 drops
    # the surface resistance (rs = 0).
    alpha = 0.05  # Albedo for wet bare soil (p. 43)
    Rns = (1 - alpha) * Rs
    Rn = Rns - Rnl
    T1 = 0.408 * Del * (Rn - G)
    T3 = Del + gam
    E0 = (T1 + T2) / T3

    return ETo, E0
